pkg/mkverification-raw-efi/vmcap.go

```diff
@@ -0,0 +1,387 @@
+// Copyright (c) 2023 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package main
+
+import (
+	"encoding/binary"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+)
+
+// For more information chekout out https://man7.org/linux/man-pages/man4/msr.4.html
+// and https://cdrdv2.intel.com/v1/dl/getContent/671200 (Appending A -VMX capability
+// reporting facility, on page 4483)
+
+const (
+	basic             = 0x480
+	pinBasedCtls      = 0x481
+	procBasedCtls     = 0x482
+	exitCtls          = 0x483
+	entryCtls         = 0x484
+	miscCtls          = 0x485
+	procBasedCtls2    = 0x48B
+	eptVPIDCap        = 0x48C
+	truePinBasedCtls  = 0x48D
+	trueProcBasedCtls = 0x48E
+	trueExitCtls      = 0x48F
+	trueEntryCtls     = 0x490
+	vmFunc            = 0x491
+	procBasedCtls3    = 0x492
+)
+
+type msr struct {
+	f *os.File
+}
+
+func newMsr() (*msr, error) {
+	f, err := os.OpenFile("/dev/cpu/0/msr", os.O_RDONLY, 0)
+	if err != nil {
+		f, err = os.OpenFile("/dev/msr0", os.O_RDONLY, 0)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &msr{f: f}, nil
+}
+
+func (m *msr) read(index uint64, defaultVal uint64) (uint64, error) {
+	p := make([]byte, 8)
+	_, err := m.f.Seek(int64(index), 0)
+	if err != nil {
+		return defaultVal, err
+	}
+	_, err = m.f.Read(p)
+	if err != nil {
+		return defaultVal, err
+	}
+
+	data := binary.LittleEndian.Uint64(p)
+
+	return data, nil
+}
+
+type control struct {
+	name    string
+	bits    map[int]string
+	msr     uint64
+	trueMsr uint64
+}
+
+func (c *control) read(nr uint64) (uint32, uint32, error) {
+	m, err := newMsr()
+	if err != nil {
+		return 0, 0, err
+	}
+	val, _ := m.read(nr, 0)
+	return uint32(val & 0xffffffff), uint32(val >> 32), err
+}
+
+func (c *control) show() {
+	fmt.Println(c.name)
+	mb1, cb1, err := c.read(c.msr)
+	tmb1, tcb1 := uint32(0), uint32(0)
+	if c.trueMsr != 0 {
+		tmb1, tcb1, err = c.read(c.trueMsr)
+	}
+	if err != nil {
+		fmt.Printf("Architecture not supported")
+		return
+	}
+	for bit := range c.bits {
+		zero := !(mb1&(1<<bit) != 0)
+		one := cb1&(1<<bit) != 0
+		trueZero := !(tmb1&(1<<bit) != 0)
+		trueOne := tcb1&(1<<bit) != 0
+		s := "?"
+		if c.trueMsr != 0 && trueZero && trueOne && one && !zero {
+			s = "default"
+		} else if zero && !one {
+			s = "no"
+		} else if one && !zero {
+			s = "forced"
+		} else if one && zero {
+			s = "yes"
+		}
+		fmt.Printf("  %-40s %s\n", c.bits[bit], s)
+	}
+}
+
+type allowedControl struct {
+	*control
+}
+
+func (a *allowedControl) read(nr uint64) (int, uint64, error) {
+	m, err := newMsr()
+	if err != nil {
+		return 0, 0, err
+	}
+	val, _ := m.read(nr, 0)
+	return 0, val, err
+}
+
+type misc struct {
+	name string
+	bits map[string]string
+	msr  uint64
+}
+
+func (m *misc) show() {
+	fmt.Println(m.name)
+	mymsr, err := newMsr()
+	if err != nil {
+		fmt.Printf("Architecture not supported")
```

Comment:
Just a final touch, the error can be also due to permissions, so it might also happen on x86 (supported architecture), so I would chose a more generic message here: "Error accessing msr registers" or "Cannot read msr registers", something like that...